File: electron/providers/tools/bash.ts
/**
 * Bash Tool - Execute shell commands
 */

import { spawn } from "child_process";

export interface BashInput {
  command: string;
  timeout?: number;
}

export interface BashOutput {
  stdout: string;
  stderr: string;
  exitCode: number;
}

const DEFAULT_TIMEOUT = 120000; // 2 minutes

export async function bashTool(input: BashInput, cwd: string): Promise<BashOutput> {
  const timeout = input.timeout ?? DEFAULT_TIMEOUT;

  return new Promise((resolve, reject) => {
    const isWindows = process.platform === "win32";
    const shell = isWindows ? "cmd.exe" : "/bin/bash";
    const shellArg = isWindows ? "/c" : "-c";

    const child = spawn(shell, [shellArg, input.command], {
      cwd,
      stdio: ["ignore", "pipe", "pipe"],
      windowsHide: true,
    });

    const stdoutChunks: Buffer[] = [];
    const stderrChunks: Buffer[] = [];

    child.stdout?.on("data", (chunk: Buffer) => {
      stdoutChunks.push(chunk);
    });

    child.stderr?.on("data", (chunk: Buffer) => {
      stderrChunks.push(chunk);
    });

    const timeoutId = setTimeout(() => {
      child.kill("SIGKILL");
      reject(new Error(`Command timed out after ${timeout}ms`));
    }, timeout);

    child.on("error", (err) => {
      clearTimeout(timeoutId);
      reject(err);
    });

    child.on("exit", (code) => {
      clearTimeout(timeoutId);
      resolve({
        stdout: Buffer.concat(stdoutChunks).toString("utf-8"),
        stderr: Buffer.concat(stderrChunks).toString("utf-8"),
        exitCode: code ?? 1,
      });
    });
  });
}

File: src-tauri/src/providers/tools/bash.rs
use anyhow::Result;
use tokio::process::Command;

/// Default timeout: 2 minutes.
const DEFAULT_TIMEOUT_MS: u64 = 120_000;

/// Result of executing a bash command.
pub struct BashResult {
    pub exit_code: i32,
    pub stdout: String,
    pub stderr: String,
}

/// Execute a shell command asynchronously.
///
/// * `command` - The shell command string to execute.
/// * `cwd` - Current working directory for the command.
/// * `timeout` - Optional timeout in milliseconds (default: 120000).
///
/// Returns `BashResult` with exit code, stdout, and stderr.
pub async fn bash_tool(
    command: &str,
    cwd: &str,
    timeout: Option<u64>,
) -> Result<BashResult> {
    let timeout_ms = timeout.unwrap_or(DEFAULT_TIMEOUT_MS);

    let child = Command::new("/bin/bash")
        .arg("-c")
        .arg(command)
        .current_dir(cwd)
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .stdin(std::process::Stdio::null())
        .kill_on_drop(true)
        .spawn()?;

    let result = tokio::time::timeout(
        std::time::Duration::from_millis(timeout_ms),
        child.wait_with_output(),
    )
    .await;

    match result {
        Ok(Ok(output)) => Ok(BashResult {
            exit_code: output.status.code().unwrap_or(1),
            stdout: String::from_utf8_lossy(&output.stdout).to_string(),
            stderr: String::from_utf8_lossy(&output.stderr).to_string(),
        }),
        Ok(Err(e)) => {
            anyhow::bail!("Command execution error: {}", e);
        }
        Err(_) => {
            // kill_on_drop will handle cleanup when child is dropped
            anyhow::bail!("Command timed out after {}ms", timeout_ms);
        }
    }
}
